### DISCUSSION TITLE
Deployed on Netlify, having an `+onBeforeRender` for a SPA dynamic pages leads to an error

### DISCUSSION THREAD

**Discussion by orangecoloured** [vikejs/vike#1389] [Discussion]
I have some dynamic pages, like `path/@index`. I used [this](https://github.com/AaronBeaudoin/vite-plugin-ssr-example-netlify/tree/main) repo as an example to set up my deployment. It's partially prerendered, and partially rendered on request.

`meta` for the dynamic pages contains `{ Page: { client: true } }`.
In the `+onBeforeRender` hook I run a quick data check, testing if the url with the `@item` exists so I can throw an error page in case it doesn't.
It works well, while I dev locally, but on Netlify it throws a 404 error.

Interesting thing is that the contents of the `+onBeforeRender` don't matter. I made it an empty function and it still failed. Removing the `+onBeforeRender` file fixes it, I can access these pages. That's what I did as a temporary fix.

How can I try to debug/fix this?

**Comment by phonzammi** [comment]
Hi @orangecoloured 

> I used this repo as an example to set up my deployment

It seems that you're using a very old example (vite-plugin-ssr) with deprecated API design. 

Please you should try to migrate to the latest Vike version with V1 Design. You can see [migration guide](https://vike.dev/migration/v1-design).

**Reply by orangecoloured** [reply]
From the example I only used [this](https://github.com/AaronBeaudoin/vite-plugin-ssr-example-netlify/tree/main#points-of-integration-summary) part in my app. The rest is aligned with the latest API design.

**Comment by brillout** [comment]
I wouldn't know how this could be related to Vike. I'm inclined to think it's some kind of user land quirk.

**Reply by orangecoloured** [reply]
Sorry, what do you mean by user land quirk?

I'm just seeing that a simple presence of `+onBeforeRender` hook file makes the app crash. And I would like to understand how to get to the cause of it. I'm not able to do it locally, because as I mentioned it works as intended while in dev mode.

**Reply by brillout** [reply]
You can't reproduce wihout Netlify, right? So it's clearly not a Vike bug.

Many already use Vike with Netlify. So either it's a new bug/regression on Netlify's side, or it's some kind of mistake on your part.

**Reply by orangecoloured** [reply]
> Many already use Vike with Netlify

Are there other examples of Netlify deployment setups? The only two I found are these: [1](https://github.com/AaronBeaudoin/vite-plugin-ssr-example-netlify/tree/main), [2](https://github.com/brunsten/vike-react-netlify/tree/main) and they are almost identical. And they are using only static routes, which do not cause crashes in my case. Would be great to see a more complex/complete setup.

> some kind of mistake on your part.

That's what would be great to understand. Otherwise the whole thing is a mystery.

**Reply by brillout** [reply]
👍 Let us know if get to the bottom of this. Don't hesite to reach out if you believe there is something Vike can do about it. And contribution welcome to add a more complete example to https://vike.dev/netlify-functions.

**Reply by orangecoloured** [reply]
I checked the function logs and there's this

```
Dec 28, 06:55:13 PM: 298abc61 ERROR  Error: [@brillout/vite-plugin-import-build@0.2.22][Wrong Usage] Cannot find server build. (Re-)build your app and try again. If you still get this error, then you may need to manually import the server build, see https://github.com/brillout/vite-plugin-import-build#manual-import
Dec 28, 06:55:13 PM: 298abc61 ERROR      at createErrorWithCleanStackTrace2 (/var/task/netlify/functions/index.js:569:15)
Dec 28, 06:55:13 PM: 298abc61 ERROR      at assertUsage3 (/var/task/netlify/functions/index.js:658:94)
Dec 28, 06:55:13 PM: 298abc61 ERROR      at loadServerBuild2 (/var/task/netlify/functions/index.js:970:31)
Dec 28, 06:55:13 PM: 298abc61 ERROR      at async loadImportBuild (/var/task/netlify/functions/index.js:2691:5)
Dec 28, 06:55:13 PM: 298abc61 ERROR      at async initGlobalContext (/var/task/netlify/functions/index.js:3335:26)
Dec 28, 06:55:13 PM: 298abc61 ERROR      at async renderPageAndPrepare (/var/task/netlify/functions/index.js:7019:5)
Dec 28, 06:55:13 PM: 298abc61 ERROR      at async renderPage_wrapper (/var/task/netlify/functions/index.js:6984:22)
Dec 28, 06:55:13 PM: 298abc61 ERROR      at async renderPage (/var/task/netlify/functions/index.js:7001:48)
Dec 28, 06:55:13 PM: 298abc61 ERROR      at async Runtime.handler (/var/task/netlify/functions/index.js:7362:23)
```

My publish directory is `dist/client/`

**Reply by brillout** [reply]
If you follow https://github.com/brillout/vite-plugin-import-build#manual-import, does it work then?

**Reply by orangecoloured** [reply]
What is `server.js` file?

**Reply by lourot** [reply]
> What is `server.js` file?

This is where your server code lives, for example:
* https://github.com/AaronBeaudoin/vite-plugin-ssr-example-netlify/blob/main/preview.ts
* https://github.com/brunsten/vike-react-netlify/blob/main/server/index.ts

**Reply by orangecoloured** [reply]
> This is where your server code lives

Sorry, I'm lost here. Is there a doc on this server code? What the convention is, folder/file structure. Why this is needed and what should be there. 
It's just while I was implementing my app I've never seen anything about this. On the vike website I can see [this](https://vike.dev/integration#server-side-tools), and it actually looks exactly like [this](https://github.com/AaronBeaudoin/vite-plugin-ssr-example-netlify/blob/main/functions/index.ts), which I've already added to my app.

**Reply by orangecoloured** [reply]
Alright, I went with the 3rd approach from [here](https://github.com/brillout/vite-plugin-import-build/issues/6). I added this to the netlify `functions/index.ts`

While deployment I got this warning
```
[WARNING] "../../../../dist/server/importBuild.cjs" should be marked as external for use with "require.resolve" [require-resolve-not-external]
9:06:33 AM:     node\_modules/@brillout/vite-plugin-import-build/dist/autoImporter.js:8:53:
9:06:33 AM:       8 │ ...: () => require.resolve("../../../../dist/server/importBuild.cjs")
```

And in the end it didn't work. The app crashes and I see the same error in logs.

**Reply by brillout** [reply]
That's a user land error. Please respect https://brillout.github.io/rules/#try. To be honest, you ask for a lot of help for things you could and should figure out yourself. We're happy to help, but do your homework before asking for help.

That said, we're very much aware that the DX around the server entry isn't perfect and we're working on it (see open PRs).

**Reply by orangecoloured** [reply]
The project I'm using it for is purely for my personal use and I don't like when having fun experimenting ends up with not having fun stumbling on bugs which I had no idea would be an issue based on the examples/docs that I used while aligning my code with the vike's API.

I ended up with `vike` only becasue the latest `create-preact` comes with it. I simply wanted to move my old `preact` code to the up to date setup. It's not like I was eager to try out `vike`, I just had to. Your documentation is not clear. And `vike` is clearly not ready for "production" use. And still you don't miss the opportunity to ask for sponsorship.

I understand that this is not a great comparison, but I can take `NextJS` and have everything I need with a much better docs, examples and community.

**Reply by brillout** [reply]
~~Be my guest and use Next.js instead of Vike. One toxic and narrow minded user less 👌~~

**Edit**: I don't feel replying like that anymore, and it isn't how I want nor feel like reacting in the future.

**Reply by brillout** [reply]
I share the sentiment that using a tool that doesn't seamlessly work is frustrating, and we very much aim to make Vike a seamless tool. (And, again, one of the open PR is actually about addressing the very issue of this thread.)

**Reply by orangecoloured** [reply]
All good. I'm going to look at svelte for now. I'm using NextJS on daily basis at work and want to try out something else for my personal use so not to lock myself into one setup/environment.

**Reply by brillout** [reply]
SvelteKit is quite neat indeed.
